File: persona1_java/src/cc4p1/raft/StateMachine.java
package cc4p1.raft;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.locks.ReentrantReadWriteLock;


public class StateMachine {

    private final List<LogEntry> encounters = new ArrayList<>();
    private final ReentrantReadWriteLock lock = new ReentrantReadWriteLock();

    public void apply(LogEntry entry) {
        lock.writeLock().lock();
        try {
            encounters.add(entry);
        } finally {
            lock.writeLock().unlock();
        }
    }

    public List<LogEntry> snapshot() {
        lock.readLock().lock();
        try {
            return Collections.unmodifiableList(new ArrayList<>(encounters));
        } finally {
            lock.readLock().unlock();
        }
    }

    public int size() {
        lock.readLock().lock();
        try {
            return encounters.size();
        } finally {
            lock.readLock().unlock();
        }
    }
}

File: persona2_python/src/raft/state_machine.py


import threading
from typing import List

from raft.log_entry import LogEntry


class StateMachine:
    def __init__(self):
        self._encounters: List[LogEntry] = []
        self._lock = threading.Lock()

    def apply(self, entry: LogEntry) -> None:
        with self._lock:
            self._encounters.append(entry)

    def snapshot(self) -> List[LogEntry]:
        with self._lock:
            return list(self._encounters)

    def size(self) -> int:
        with self._lock:
            return len(self._encounters)
